doc/adding_lints.md

```diff
@@ -0,0 +1,324 @@
+Running `TESTNAME=ui/foo_functions cargo uitest` should pass then. When you
+commit your lint, be sure to commit the `*.stderr` files, too.
+
+Let's have a look at implementing our lint now.
+
+### Lint declaration
+
+We start by creating a new file in the `clippy_lints` crate. That's the crate
+where all the lint code is. We are going to call the file
+`clippy_lints/src/foo_functions.rs` and import some initial things we need:
+
+```rust
+use rustc::lint::{LintArray, LintPass};
+use rustc::{declare_tool_lint, lint_array};
+```
+
+The next step is to provide a lint declaration. Lints are declared using the [`declare_clippy_lint!`][declare_clippy_lint] macro:
+
+```rust
+declare_clippy_lint! {
+    pub FOO_FUNCTIONS,
+    pedantic,
+    "function named `foo`, which is not a descriptive name"
+}
+```
+
+* `FOO_FUNCTIONS` is the name of our lint. Be sure to follow the [lint naming
+guidelines][lint_naming] here when naming your lint. In short, the name should
+state the thing that is being checked for and read well when used with
+`allow`/`warn`/`deny`.
+* `pedantic` sets the lint level to `Allow`.
+  The exact mapping can be found [here][category_level_mapping]
+* The last part should be a text that explains what exactly is wrong with the
+  code
+
+With our lint declaration done, we will now make sure that our lint declaration is assigned to a lint pass:
+
+```rust
+// clippy_lints/src/foo_functions.rs
+
+// .. imports and lint declaration ..
+
+#[derive(Copy, Clone)]
+pub struct FooFunctionsPass;
+
+impl LintPass for FooFunctionsPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(
+            FOO_FUNCTIONS,
+        )
+    }
+
+    fn name(&self) -> &'static str {
+        "FooFunctions"
+    }
+}
+```
+
+Don't worry about the `name` method here. As long as it includes the name of the lint pass it should be fine.
+
+Next you should run `util/dev update_lints` to register the lint in various
+places, mainly in `clippy_lints/src/lib.rs`.
+
+While `update_lints` automates some things, it doesn't automate everything. We will have to register our lint pass manually in the `register_plugins` function in `clippy_lints/src/lib.rs`:
+
+```rust
+reg.register_early_lint_pass(box foo_functions::FooFunctionsPass);
+```
+
+Without that, running the UI tests would produce an error like `unknown clippy lint: clippy::foo_functions`.
+The next decision we have to make is which lint pass our lint is going to need.
+
+### Lint passes
+
+Writing a lint that just checks for the name of a function means that we just
+have to deal with the AST and don't have to deal with the type system at all.
+This is good, because it makes writing this particular lint less complicated.
+
+We have to make this decision with every new Clippy lint. It boils down to using either `EarlyLintPass` or `LateLintPass`. This is a result of Rust's compilation process. You can read more about it [in the rustc guide][compilation_stages].
```

Comment:
You link to the rustc guide, but that page doesn't mention early/late. Maybe mention in which stage the early/late lints are run?

Reply:
You're right, I think I will leave that last sentence out for now. There's not really a place where EarlyLintPass and LateLintPass are explained apart from this PR.